Media: Processing items missing after opening media dialog

spacedmonkey opened the pull request:
## Context

<!-- What do we want to achieve with this PR? Why did we write this code? -->

## Summary

Add `onUpload` and `onDelete` callbacks to media upload component. This has a number of key benefits. This allow for media  added and removed from within the editor. For example, if an image is delete from the dialog, it is removed from the media library and on the canvas. 

I have also taken this opportunity to use this new `onDelete` to improve background audio and captions. If these files are deleted within the dialog, then background audio and caption are reset / removed. 

## Relevant Technical Choices

<!-- Please describe your changes. -->

## To-do

<!-- A list of things that need to be addressed in this PR or follow-up changes. -->

## User-facing changes

https://user-images.githubusercontent.com/237508/179557163-02d60356-78a7-4eb7-ac6e-9083b1fcca83.mov

## Testing Instructions

<!--
How can the changes in this PR be verified?
Please provide step-by-step instructions how to reproduce the issue, if applicable.
Write step-by-step test instructions aimed at non-tech-savvy users, even if the PR is not user-facing.
-->

<!-- ignore-task-list-start -->
- [ ] This is a non-user-facing change and requires no QA
<!-- ignore-task-list-end -->

This PR can be tested by following these steps:

1. Create a story. 
2. Open media dialog. 
3. Upload image. 
4. Insert into canvas. 
5. See new image into media library. 
---
1. Create a story. 
2. Open media dialog. 
3. Insert image into canvas. 
4. Open media dialog. 
5. Select image ( the one used before ). 
6. Delete image 
7. Close dialog. 
8. See image removed from canvas and in media library.
---
1. Create a story. 
2. Select page
3. Go to style tab. 
4. Upload audio file.
5. Select file. 
6. Click replace file. 
7. Delete image ( the one used before ). 
8. Close dialog. 
9. See background audio is reset. 

## Reviews

### Does this PR have a security-related impact?

<!-- Examples: new APIs, changes to KSES, etc.  -->

### Does this PR change what data or activity we track or use?

<!-- Examples: changes to telemetry, new third-party APIs -->

### Does this PR have a legal-related impact?

<!-- Examples: new images with unknown sources, new production dependencies with incompatible licenses -->

## Checklist

<!-- Check these after PR creation -->

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

<!--
Please reference the issue(s) this PR addresses.
No URLs, just the issue numbers.
Use "Fixes #123" if it fixes an issue.

NOTE: One reference per line!

Example:

Fixes #123
Partially addresses #456
See #789
-->

Fixes #9963

spacedmonkey committed:
Add onUpload and onDelete callbacks.

swissspidy commented on packages/story-editor/src/app/highlights/quickActions/useQuickActions.js:
```diff
@@ -108,14 +115,16 @@ export const MediaPicker = ({ render, ...props }) => {
     ({
       state: { canTranscodeResource },
       actions: {
-        resetWithFetch,
```
Can we now completely remove the `resetWithFetch` function from `packages/story-editor/src/app/media/local/useContextValueProvider.js`?

spacedmonkey replied:
Removed.

spacedmonkey committed:
Fix lints.

spacedmonkey committed:
Merge branch 'main' into fix/dontclear

swissspidy commented on packages/story-editor/src/components/form/mediaUploadButton.js:
```diff
@@ -157,11 +164,29 @@ function MediaUploadButton({ onInsert, renderButton, buttonInsertText }) {
+  const onDelete = useCallback(
+    (resource) => {
+      deleteMediaElement({ id: resource?.id });
+      deleteElementsByResourceId({ id: resource?.id });
```
Why the optional chaining, can `resource` really be `null` here?

swissspidy commented on packages/story-editor/src/app/highlights/quickActions/useQuickActions.js:
```diff
@@ -217,12 +226,31 @@ export const MediaPicker = ({ render, ...props }) => {
+  const onDelete = useCallback(
+    (resource) => {
+      deleteMediaElement({ id: resource?.id });
+      deleteElementsByResourceId({ id: resource?.id });
```
Why the optional chaining, can `resource` really be `null` here?

swissspidy commented on packages/story-editor/src/components/panels/shared/media/captionsPanelContent.js:
```diff
@@ -104,6 +104,12 @@ function CaptionsPanelContent({
   const onError = useCallback((err) => trackError(eventName, err?.message), []);
+  const onDeleteTrack = useCallback(
+    (deletedResource) => {
+      handleRemoveTrack(deletedResource?.id);
```
Why the optional chaining, can `deletedResource ` really be `null` here?

swissspidy commented on packages/wp-story-editor/src/components/mediaUpload/mediaPicker/useMediaPicker.js:
```diff
@@ -92,7 +99,58 @@ function useMediaPicker({
+          });
+
+          this.moveFocus();
+        }
```
Where is this copied from? Instead of copying, can we just call the original `deleteAttachment` function or something? Ditto for `trashAttachment` below

spacedmonkey replied:
Fixed.

spacedmonkey committed:
Remove ResetWithFetch.

spacedmonkey committed:
No chains on me.

spacedmonkey committed:
Call the parent.

spacedmonkey committed:
No more chains.

spacedmonkey committed:
Also reset properties

spacedmonkey committed:
Merge branch 'main' into fix/dontclear

swissspidy commented on packages/wp-story-editor/src/components/mediaUpload/mediaPicker/useMediaPicker.js:
```diff
@@ -92,7 +99,35 @@ function useMediaPicker({
+        }
+      },
+    });
+    window.wp.media.view.Attachment.Details = wsDetails;
```
Probably want to do a cleanup here:

```suggestion
    window.wp.media.view.Attachment.Details = wsDetails;
    
    return () => {
      window.wp.media.view.Attachment.Details = currentDetails;
    };
```

swissspidy commented on packages/story-editor/src/app/highlights/quickActions/useQuickActions.js:
```diff
@@ -217,12 +226,31 @@ export const MediaPicker = ({ render, ...props }) => {
+    (resource) => {
+      prependMedia({
+        media: [resource],
+      });
```
What if I filtered my media library by just images or videos instead of "All Types"? Then `prependMedia()` causes a video to be prepended to my list of images like this:

![Screenshot 2022-07-19 at 13 10 01](https://user-images.githubusercontent.com/841956/179736621-dbec9fec-206a-4f61-9350-f3a2a8b516c8.png)

I see two options here:

1. We wrap this call in a check for `mediaType`
2. We add a `mediaType` check directly to the `prependMedia` reducer

Option (2) seems more robust, in case we forgot to do (1) in some places. Plus we can easily add tests.

WDYT?

spacedmonkey replied:
I am seeing this issue in main. So if a drag a video into canvas when I am filtering image, I see it prepended to media library. I think We should add this to `prependMedia` reducer but in maybe into another PR, as it is another issue,

spacedmonkey replied:
This took a lot longer to track down than I thought. I have a fix here, see what you https://github.com/GoogleForCreators/web-stories-wp/pull/11941/commits/dbfeb95d7c03162512628453214e9916c5a31af5

swissspidy replied:
> I am seeing this issue in main. So if a drag a video into canvas when I am filtering image, I see it prepended to media library. 

Right. Just found #9964 because of this, indicating that we did that on purpose.

dbfeb95d7c03162512628453214e9916c5a31af5 means that you'll still see the uploading items when you change the filter again, right?

That couldn't be done in the reducer, so it kinda makes sense to put it there.

On the other hand, it's weird that we have to filter the media elements at that stage. Would be better to put this into `useContextValueProvider` directly, not have it in the component.

But yes, let's do that in a separate PR so we can refine & test it better.

spacedmonkey replied:
I have reverted https://github.com/GoogleForCreators/web-stories-wp/commit/dbfeb95d7c03162512628453214e9916c5a31af5 let's handle this in https://github.com/GoogleForCreators/web-stories-wp/issues/9964.

spacedmonkey replied:
Left my notes on this here https://github.com/GoogleForCreators/web-stories-wp/issues/9964#issuecomment-1190102837

swissspidy commented on packages/story-editor/src/components/form/mediaUploadButton.js:
```diff
@@ -157,11 +164,29 @@ function MediaUploadButton({ onInsert, renderButton, buttonInsertText }) {
+      deleteElementsByResourceId({ id: resource.id });
+    },
+    [deleteElementsByResourceId, deleteMediaElement]
+  );
```
Since this is identical to the code in `useQuickActions.js`, and overall the pattern is quite repetitive, I am wondering whether it would be easiest to move (parts of) this to a shared component.

spacedmonkey replied:
Fixed.

spacedmonkey replied:
Fixed. Reused compontent.

spacedmonkey committed:
Apply suggestions from code review

Co-authored-by: Pascal Birchler <pascalb@google.com>

spacedmonkey committed:
Don't repeat yourself.

spacedmonkey committed:
Fix lint.

miina commented on packages/story-editor/src/components/publishModal/content/storyPreview.js:
```diff
@@ -298,6 +311,7 @@ const StoryPreview = () => {
                 {(hasUploadMediaAction || enablePosterHotlinking) && (
                   <StyledMedia
                     onChange={handleChangePoster}
+                    onDelete={onDelete}
```
Just curious, wasn't deleting purposefully not added to the publishing modal? Since it's required for publishing.

spacedmonkey replied:
https://github.com/GoogleForCreators/web-stories-wp/pull/11941#discussion_r925324309

miina commented on packages/story-editor/src/components/panels/design/videoAccessibility/videoAccessibility.js:
```diff
@@ -171,6 +180,7 @@ function VideoAccessibilityPanel({ selectedElements, pushUpdate }) {
           value={poster}
           cropParams={cropParams}
           onChange={handleChangePoster}
+          onDelete={onDelete}
```
Here the Delete is essentially the same as Reset now, right? Since it's not actually fully deleted for the user, there's always a default. Or in case of Delete, the image is actually deleted from the library as well, in case of Reset not?

miina replied:
^ I wouldn't think as a user that "Delete" here would mean deleting from the library as well 🤔 I'd assume that the poster would just be deleted from that field.

spacedmonkey replied:
@miina I don't think you understand what delete means in this context. If a user clicks delete in this dialog, it is completely deleted, file and all. 

<img width="284" alt="Screenshot 2022-07-20 at 09 26 58" src="https://user-images.githubusercontent.com/237508/179934952-8e428455-924b-4a38-a9b3-652a90defe4c.png">

In that context, reseting makes sense here, as the file the user selected no longer exists.

swissspidy replied:
Is it worth renaming `onDelete` to something else to make it less confusing?

spacedmonkey replied:
How about onDeleteMedia and onUploadMedia?

miina replied:
Yep, that is more clear when reading the code.

spacedmonkey replied:
Changed.

swissspidy commented on packages/story-editor/src/app/highlights/quickActions/useQuickActions.js:
```diff
@@ -180,51 +107,11 @@ export const MediaPicker = ({ render, ...props }) => {
-  );
   return (
-    <MediaUpload
+    <MediaUploadButton
```
Neat, always nice to see code removals like this. And it works without any issues?

spacedmonkey replied:
Seems to.

swissspidy commented on packages/story-editor/src/app/media/local/reducers/setMediaType.js:
```diff
@@ -24,7 +28,7 @@ function setMediaType(state, { mediaType }) {
   return {
     ...INITIAL_STATE,
-    media: state.media.filter(({ local }) => local), // This filter allows remove temporary file returned on upload
+    media: state.media.filter(({ src }) => isBlobURL(src)), // This filter allows remove temporary file returned on upload
```
Not sure I really understand the purpose of this filter here, even in it's previous form 🤔 

Doesn't that _keep_ temporary files instead of _removing_ them?

spacedmonkey replied:
Before, we had a local property on the resource object. This broke a while back. This is a check to filter out all "local" resources. We can no longer do this as state is not managed here anymore.

swissspidy replied:
Yeah but to filter out local resources you'd do `state.media.filter(({ local }) => ! local)` and not `state.media.filter(({ local }) => local)` 

Again, this would _keep_ the items, not _remove_ them, which seems odd.

Am I missing something?

spacedmonkey committed:
Move upstream.

spacedmonkey committed:
Merge branch 'main' into fix/dontclear

spacedmonkey committed:
Unused variable.

spacedmonkey committed:
Merge branch 'main' into fix/dontclear

spacedmonkey committed:
Change param name.

spacedmonkey committed:
Change param name again.

swissspidy commented on packages/wp-story-editor/src/components/mediaUpload/mediaPicker/useMediaPicker.js:
```diff
@@ -75,6 +84,44 @@ function useMediaPicker({
+      }
+    },
+    [deleteElementsByResourceId, deleteMediaElement, onDeleteMedia]
+  );
```
This should not be in the wp-story-editor package, since actions like prepending media and deleting media elements is something that also needs to happen in a decoupled/standalone context.

spacedmonkey committed:
Move back to story-editor.

spacedmonkey commented on packages/story-editor/src/components/mediaUpload/mediaUpload.js:
```diff
@@ -0,0 +1,104 @@
+        prependMedia({
+          media: [resource],
+        });
+      }
```
We could call `postProcessingResource` to trigger post processing resource early.

swissspidy replied:
Doesn't seem necessary because it will be called right away already due to this effect:

https://github.com/GoogleForCreators/web-stories-wp/blob/22cf54d030e407f469350f21273f6d1c223aa99b/packages/story-editor/src/app/media/local/useContextValueProvider.js#L288-L292

spacedmonkey committed:
Remove on Delete.

spacedmonkey committed:
Check if callback exists.

spacedmonkey committed:
Change param name again.

spacedmonkey committed:
Dry.

spacedmonkey committed:
Merge branch 'main' into fix/dontclear

spacedmonkey committed:
Lints.

spacedmonkey committed:
Merge branch 'main' into fix/dontclear

spacedmonkey committed:
Refactor to use queue.

timarney commented on packages/story-editor/src/components/mediaUpload/mediaUpload.js:
```diff
@@ -0,0 +1,129 @@
+      deleteMediaElement({ id: resource.id });
+      deleteElementsByResourceId({ id: resource.id });
+      setNewResources((state) => state.filter(({ id }) => id !== resource.id));
+      resetFeatureMedia(resource.id);
```
```suggestion
      const id = resource.id;
      deleteMediaElement({ id });
      deleteElementsByResourceId({ id });
      setNewResources((state) => state.filter(({ newId }) => newId !== id));
      resetFeatureMedia(id);
```

swissspidy committed:
Merge branch 'main' into fix/dontclear

swissspidy committed:
Fix prop passed to `MediaUploadButton`

Prevents prop type warning in Quick Actions menu

swissspidy committed:
Do not store `fileFrame` in ref to avoid stale data

Otherwise, when deleting an item in the modal and closing & reopening it, you’d still see the deleted item there.

swissspidy committed:
Remove outdated comment

swissspidy committed:
Use ref for `onSelect`

swissspidy committed:
Add todo
